devlib/connection.py

```diff
@@ -72,6 +72,38 @@ def __del__(self):
         if self.initialized:
             self.close()
 
+    def _expand_glob(self, pattern):
+        """
+        Expand the given path globbing pattern on the target using the shell
+        globbing.
+        """
+        # Since we split the results based on new lines, forbid them in the
+        # pattern
+        if '\n' in pattern:
+            raise ValueError(r'Newline character \n are not allowed in globbing patterns')
+
+        # Characters to escape that are impacting parameter splitting, since we
+        # want the pattern to be given in one piece. Unfortunately, there is no
+        # fool-proof way of doing that without also escaping globbing special
+        # characters such as wildcard which would defeat the entire purpose of
+        # that function.
+        for c in [' ', "'", '"']:
+            pattern = pattern.replace(c, '\\' + c)
+
+        cmd = "printf '%s\n' {}".format(pattern)
+        result = self.execute(cmd, strip_colors=False)
+        paths = result.splitlines()
+        if not paths:
+            raise TargetStableError('No file matching: {}'.format(pattern))
```

Comment:
Even when there are no matches will `printf` not echo the pattern itself back?

Reply:
If there is no match, printf won't be called. What happens is shell dependent: on most shells, $? will be set to non-zero value and the shell itself will print an error message saying there was no match. On busybox ash, the command won't be called and $?=0. On ZSH with the right option enabled, it can also decide to just ignore the non-matching pattern and just skip that parameter entirely.

AFAIK, no shell will ever call the command with the pattern itself though if there is no match.